### DISCUSSION TITLE
Multiple border routers causing all nodes not to be able to send messages to the internet

### DISCUSSION THREAD

**Discussion by meatherly** [openthread/openthread#8451] [Q&A] [answered]
Hello,

The issue I'm running into is that whenever a 2nd border router is connected to the network no device can send requests outside the network to the internet. I'm not sure what is causing this to happen. The only thing I do notice is in the logs whenever 2nd border router connects I start to get a lot of Checksum errors.  

```
I (601918) OPENTHREAD: Received PIO
I(605498) OPENTHREAD:[N] Ip6-----------: Error UDP Checksum
I(605498) OPENTHREAD:[N] Ip6-----------: Failed to handle payload: Drop
I(626888) OPENTHREAD:[N] Ip6-----------: Error UDP Checksum
I(626888) OPENTHREAD:[N] Ip6-----------: Failed to handle payload: Drop
```

I'm not sure if this is causing the issue or not though. Whenever I only have 1 border router on the network devices have no problem sending requests to the internet. Strangely enough whenever I remove the 2nd border router from the network nodes still cannot connect to the internet. It's like the network is left in a non-recoverable state. I have to power down the end node, remove the 2nd border router, factory reset the 1st border router, and then power up the end node. After all that the node can connect to the internet.

My Setup:

* 2 border routers using ESP32s and Nordic 52840 Dongles for the RCP. [openthread versions](https://github.com/espressif/esp-idf/tree/release/v5.0/components/openthread). They are both running this [example code](https://github.com/espressif/esp-idf/blob/release/v5.0/examples/openthread/ot_br/main/esp_ot_br.c) 
* 1 Nordic 52840 Dongle as a FTD node

I've predefined all the network credentials on all nodes using KConfigs:
```
#
# OpenThread
#
CONFIG_OPENTHREAD_NETWORK_NAME="OpenThread"
CONFIG_OPENTHREAD_NETWORK_CHANNEL=15
CONFIG_OPENTHREAD_NETWORK_PANID=0x1234
CONFIG_OPENTHREAD_NETWORK_EXTPANID="dead00beef00cafe"
CONFIG_OPENTHREAD_NETWORK_MASTERKEY="00112233445566778899aabbccddeeff"
CONFIG_OPENTHREAD_NETWORK_PSKC="104810e2315100afd6bc9215a6bfac53"
# end of OpenThread
```

The 3rd node outputs this in the logs: 
```
[00:01:56.494,720] <inf> [I] Mle-----------: Attempt to become router
[00:01:56.495,483] <inf> [I] Mle-----------: Send Address Solicit (fdd3:9184:27d5:b53b:0:ff:fe00:7400)
[00:01:56.500,183] <inf> [I] MeshForwarder-: Sent IPv6 UDP msg, len:77, chksum:be22, ecn:no, to:0x7400, sec:yes, prio:net
[00:01:56.500,488] <inf> [I] MeshForwarder-:     src:[fdd3:9184:27d5:b53b:0:ff:fe00:7402]:61631
[00:01:56.500,823] <inf> [I] MeshForwarder-:     dst:[fdd3:9184:27d5:b53b:0:ff:fe00:7400]:61631
[00:01:59.481,567] <inf> [I] MeshForwarder-: Sent IPv6 UDP msg, len:77, chksum:be22, ecn:no, to:0x7400, sec:yes, prio:net
[00:01:59.481,903] <inf> [I] MeshForwarder-:     src:[fdd3:9184:27d5:b53b:0:ff:fe00:7402]:61631
[00:01:59.482,238] <inf> [I] MeshForwarder-:     dst:[fdd3:9184:27d5:b53b:0:ff:fe00:7400]:61631
[00:02:05.446,838] <inf> [I] MeshForwarder-: Sent IPv6 UDP msg, len:77, chksum:be22, ecn:no, to:0x7400, sec:yes, prio:net
[00:02:05.447,204] <inf> [I] MeshForwarder-:     src:[fdd3:9184:27d5:b53b:0:ff:fe00:7402]:61631
[00:02:05.447,509] <inf> [I] MeshForwarder-:     dst:[fdd3:9184:27d5:b53b:0:ff:fe00:7400]:61631
[00:02:13.610,717] <dbg> [D] Mac-----------: Frame rx failed, error:DestinationAddressFiltered
[00:02:17.375,152] <inf> [I] MeshForwarder-: Sent IPv6 UDP msg, len:77, chksum:be22, ecn:no, to:0x7400, sec:yes, prio:net
[00:02:17.375,488] <inf> [I] MeshForwarder-:     src:[fdd3:9184:27d5:b53b:0:ff:fe00:7402]:61631
[00:02:17.375,793] <inf> [I] MeshForwarder-:     dst:[fdd3:9184:27d5:b53b:0:ff:fe00:7400]:61631
[00:02:24.377,471] <inf> [I] MeshForwarder-: Received IPv6 UDP msg, len:90, chksum:5926, ecn:no, from:ea82c09ea0821acb, sec:no, prio0
[00:02:24.377,807] <inf> [I] MeshForwarder-:     src:[fe80:0:0:0:e882:c09e:a082:1acb]:19788
[00:02:24.378,143] <inf> [I] MeshForwarder-:     dst:[ff02:0:0:0:0:0:0:1]:19788
[00:02:24.379,089] <inf> [I] Mle-----------: Receive Advertisement (fe80:0:0:0:e882:c09e:a082:1acb,0x7400)
```

Please let me know if I need to provide any more information to help solve this issue.

**Comment by jwhui** [comment]
I believe the issue is because the Active Operational Dataset is not matching across your devices. In particular, because you did not specify the Mesh Local Prefix in your configuration, the devices are not able to communicate properly in the Thread network.

I suggest the following the OpenThread Border Router codelab. Specifically:
1. [Form a Thread network](https://openthread.io/codelabs/openthread-border-router#2)
2. [Join the OTBR network](https://openthread.io/codelabs/openthread-border-router#3)

**Reply by meatherly** [reply]
Thank you for your response. I have now added the same mesh local prefix to all my devices. This has stopped the UDP checksum errors. I'm still having trouble getting nodes to reach the internet whenever there are 2 border routers on the network. I thought OpenThread supports multiple border routers? 

Here's some logs to help debug this issue
Node sending ping:
```
uart:~$ ot ping 8.8.8.8
Pinging synthesized IPv6 address: fdae:a713:fe2b:2:0:0:808:808
[00:03:45.245,025] <inf> [I] Icmp6---------: Sent echo request: (seq = 2)
[00:03:45.249,725] <inf> [I] MeshForwarder-: Sent IPv6 ICMP6 msg, len:56, chksum:f842, ecn:no, to:0x2000, sec:yes, prio:normal
[00:03:45.250,061] <inf> [I] MeshForwarder-:     src:[fdf8:2e54:ed95:1:e895:72be:d4a9:1aac]
[00:03:45.250,366] <inf> [I] MeshForwarder-:     dst:[fdae:a713:fe2b:2:0:0:808:808]
1 packets transmitted, 0 packets received. Packet loss = 100.0%.
Done
```

Border router logs:
```
I(492429) OPENTHREAD:[I] MeshForwarder-: Received IPv6 ICMP6 msg, len:56, chksum:f842, ecn:no, from:0xa000, sec:yes, prio:normal, rss:-51.0
I(492429) OPENTHREAD:[I] MeshForwarder-:     src:[fdf8:2e54:ed95:1:e895:72be:d4a9:1aac]
I(492439) OPENTHREAD:[I] MeshForwarder-:     dst:[fdae:a713:fe2b:2:0:0:808:808]
```
So the border router gets the request but then nothing happens. 


But when there is only 1 border router on the network things work.

Logs from the node sending the ping
```
uart:~$ ot ping 8.8.8.8
Pinging synthesized IPv6 address: fd8a:993c:b044:2:0:0:808:808
16 bytes from fd8a:993c:b044:2:0:0:808:808: icmp_seq=1 hlim=53 time=88ms
1 packets transmitted, 1 packets received. Packet loss = 0.0%. Round-trip min/avg/max = 88/88.0/88 ms.
Done
```

Logs from the border router:
```
I(33174) OPENTHREAD:[I] MeshForwarder-: Received IPv6 ICMP6 msg, len:56, chksum:8ed7, ecn:no, from:0x6c01, sec:yes, prio:normal, rss:-43.0
I(33174) OPENTHREAD:[I] MeshForwarder-:     src:[fd8a:993c:b044:1:3db2:7a69:646b:edb6]
I(33194) OPENTHREAD:[I] MeshForwarder-:     dst:[fd8a:993c:b044:2:0:0:808:808]
I(33244) OPENTHREAD:[I] MeshForwarder-: Sent IPv6 ICMP6 msg, len:56, chksum:8dd7, ecn:no, to:0x6c01, sec:yes, prio:normal
I(33254) OPENTHREAD:[I] MeshForwarder-:     src:[fd8a:993c:b044:2:0:0:808:808]
I(33264) OPENTHREAD:[I] MeshForwarder-:     dst:[fd8a:993c:b044:1:3db2:7a69:646b:edb6]
```

**Reply by jwhui** [reply]
The Thread device seems to switch to a different NAT64 prefix when the second BR is added. I assume that new NAT64 prefix is coming from the second BR. I'm curious, does NAT64 work when the second BR is the only BR in the network? I'm wondering if there is an issue specific to the second BR.

**Reply by meatherly** [reply]
Interesting. 
If ask for the `netdata` it shows this when there is 2 border routers:
```
> netdata show
Prefixes:
fd1d:2813:9ce5:1::/64 paos low 4c00
Routes:
2600:1700:4fe0:ee2f::/64 s med 4c00
2600:1700:4fe0:ee2f::/64 s med 2000
fd1d:2813:9ce5:2:0:0::/96 sn low 4c00
fd4f:68a:de46:1::/64 s med 4c00
fd4f:68a:de46:1::/64 s med 2000
Services:
44970 01 70000500000e10 s 4c00
44970 5d fd1ef509f59cf14baf93a5e3f2d35addd11f s 4c00
44970 5d fd1ef509f59cf14b1f294e78ef370df0d11f s 2000
Done
```
Is this normal?

**Reply by meatherly** [reply]
Just let me know what information you need to hunt this down I'll do my best to get it for you!

**Reply by jwhui** [reply]
The NAT64 prefix is marked by the `n` flag. The network data output above shows only one NAT64 prefix:
```
fd1d:2813:9ce5:2:0:0::/96 sn low 4c00
```
And the NAT64 prefix is published by the BR with RLOC16 = 0x4c00.

This all looks fine to me. When you try to send to 8.8.8.8, is it using the above NAT64 prefix? If NAT64 is enabled on 0x4c00, this should work as expected.

**Reply by meatherly** [reply]
Hey @jwhui 
Yeah it sends the message to the right prefix and the BR that registered that prefix receives the messages but then I get a timeout on the end node that made the request because the border router never replied back. Is there a callback I can register on the BR NAT64 API to see what's happening on the BR? Do you know of any log APIs I should turn on in the ESP to get better information on what could be happening?

**Reply by jwhui** [reply]
Unfortunately, I don't have much direct experience with Espressif's border router implementation. Maybe @chshu can provide some insight.

**Comment by zwx1995esp** [comment]
@meatherly Hi, I will try to reproduce this issue, and do some investigations.

**Comment by zwx1995esp** [comment, marked as answer]
@meatherly Hi, I have already found the reason why end device can not send the message to the internet, when multi-BRs join in the OT network. If two BRs join in a same openthread partition, on the ESP BR implementation, miss a logic for checking whether the packet received on wifi network need forward to openthread network. The fixing MR has already created in our internal branch. This bug will be also fixed on the esp-idf master branch on the github, but still takes some time. If you want to use this feature, I can build a lib for you right now.

**Reply by meatherly** [reply]
Thank you so much for getting back to me on this. What would be the timeline for the fix if you don't mind me asking? Maybe sometime in the next couple of weeks? 

I did make an issue on ESP-IDF if you wanted to reference it: https://github.com/espressif/esp-idf/issues/10291

**Reply by zwx1995esp** [reply]
Actually, the fix MR now is running CI on our internal branch. For the time you can get the fix on the github master branch, maybe next week(if all things go as expected).

**Reply by chshu** [reply]
@meatherly The fix has been synced to github, could you test again with the latest IDF master branch?
